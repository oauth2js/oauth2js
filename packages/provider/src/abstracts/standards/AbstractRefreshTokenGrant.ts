import { Mixed } from '@prunus/mixin'

import { InvalidClient } from '../../errors/InvalidClient'
import { InvalidGrant } from '../../errors/InvalidGrant'
import {
  DEFAULT_ACCESS_TOKEN_LIFETIME,
  DEFAULT_REFRESH_TOKEN_LIFETIME,
} from '../../helpers/defaults'
import { Maybe, PromiseToo } from '../../helpers/types'
import { RequestLocation } from '../../interfaces/IRequest'
import { IToken } from '../../interfaces/IToken'
import { Context } from '../../models/Context'
import { UseClientId } from '../../traits/UseClientId'
import { UseTokenByRefreshToken } from '../../traits/UseTokenByRefreshToken'
import { AbstractGrant } from '../AbstractGrant'

export abstract class AbstractRefreshTokenGrant extends Mixed(
  AbstractGrant,
  UseTokenByRefreshToken,
  UseClientId,
) {
  protected accessTokenLifetime: number = DEFAULT_ACCESS_TOKEN_LIFETIME
  protected refreshTokenLifetime: number = DEFAULT_REFRESH_TOKEN_LIFETIME
  protected readonly clientIdLocation: RequestLocation = RequestLocation.Body

  protected readonly refreshTokenLocation: RequestLocation =
    RequestLocation.Body

  protected async createToken(context: Context): Promise<IToken> {
    const token = await this.getOrLoadTokenByRefreshToken(context)

    if ((await this.getClientId(context)) !== token.client.id)
      throw new InvalidClient('Client identifier does not match', context)

    const { client, user, scopes } = token
    const [
      accessToken,
      accessTokenExpiresAt,
      refreshToken,
      refreshTokenExpiresAt,
    ] = await Promise.all([
      this.generateAccessToken(context),
      this.getAccessTokenExpiresAt(context),
      this.generateRefreshToken(context),
      this.getRefreshTokenExpiresAt(context),
    ])

    return {
      client,
      user,
      scopes,
      accessToken,
      accessTokenExpiresAt,
      refreshToken,
      refreshTokenExpiresAt,
    }
  }

  protected async saveToken(token: IToken, context: Context): Promise<IToken> {
    return await this.saveTokenByRefreshedToken(
      token,
      await this._revokeToken(context),
      context,
    )
  }

  private async _revokeToken(context: Context) {
    const tokenToRevoke = await this.getOrLoadTokenByRefreshToken(context)

    if (!(await this.revokeToken(tokenToRevoke, context)))
      throw new InvalidGrant('The token is invalid', context)

    return tokenToRevoke
  }

  protected abstract override getTokenByRefreshToken(
    refreshToken: string,
    context: Context,
  ): PromiseToo<Maybe<IToken>>

  protected abstract saveTokenByRefreshedToken(
    token: IToken,
    refreshedToken: IToken,
    context: Context,
  ): PromiseToo<IToken>

  protected abstract revokeToken(
    token: IToken,
    context: Context,
  ): PromiseToo<boolean>
}
