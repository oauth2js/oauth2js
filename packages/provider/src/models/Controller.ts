import { is } from 'type-is'

import { AbstractExchanger } from '../abstracts/AbstractExchanger'
import { AbstractFlow } from '../abstracts/AbstractFlow'
import { AbstractGrant } from '../abstracts/AbstractGrant'
import { AbstractTokenType } from '../abstracts/AbstractTokenType'
import { isOAuth2Error } from '../abstracts/OAuth2Error'
import { InvalidRequest } from '../errors/InvalidRequest'
import { ServerError } from '../errors/ServerError'
import { UnsupportedGrantType } from '../errors/UnsupportedGrantType'
import { GRANT_TYPE_REGEXP, RESPONSE_TYPE_REGEXP } from '../helpers/regexps'
import {
  ArrayInfer,
  ArrayReadOnlyToo,
  ConstructorOf,
  Maybe,
} from '../helpers/types'
import { IMPLEMENTATION_TYPE } from '../interfaces/IControllerImplementation'
import { IRequest } from '../interfaces/IRequest'
import { IResponse } from '../interfaces/IResponse'
import { ITicket } from '../interfaces/ITicket'
import { IUser } from '../interfaces/IUser'
import { TokenBearerType } from '../standards/TokenBearerType'
import { Context } from './Context'

export type Index = [
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  16,
  17,
  18,
  19,
  20,
  -1,
]

export type WalkAnd<
  T extends ArrayReadOnlyToo<ConstructorOf<unknown>>,
  K extends string | symbol,
  I extends number = 0,
> = I extends -1
  ? never
  : T[I] extends undefined
  ? never
  : K extends keyof InstanceType<T[I]>
  ? T[Index[I]] extends undefined
    ? InstanceType<T[I]>[K]
    : InstanceType<T[I]>[K] | WalkAnd<T, K, Index[I]>
  : never

export type ImplementationOf<
  G extends
    | ArrayReadOnlyToo<ConstructorOf<AbstractGrant>>
    | ConstructorOf<AbstractGrant>,
> = G extends ArrayReadOnlyToo<ConstructorOf<AbstractGrant>>
  ? Partial<WalkAnd<G, IMPLEMENTATION_TYPE>>
  : G extends ConstructorOf<AbstractGrant>
  ? Partial<WalkAnd<[G], IMPLEMENTATION_TYPE>>
  : never

export class Controller<
  G extends ArrayReadOnlyToo<ConstructorOf<AbstractGrant>> = [],
> {
  private _grants: AbstractGrant[] = []
  private _flows: AbstractFlow[] = []
  private _tokenTypes: AbstractTokenType[] = [new TokenBearerType()]

  public use<T extends ConstructorOf<AbstractGrant>>(
    Grant: T,
  ): Controller<[...G, T]> {
    const grant = new Grant()

    this._grants.push(grant)

    grant['whenBeingAttached'](this as Controller)

    if (grant instanceof AbstractExchanger && 'flow' in grant) {
      const flow = grant['flow']

      this._flows.push(flow)

      flow['whenBeingAttached'](this as Controller)
    }

    return this as unknown as Controller<[...G, T]>
  }

  public implement(implementations: ImplementationOf<G>): this
  public implement<T extends ArrayReadOnlyToo<ArrayInfer<G>>>(
    implementations: ImplementationOf<T>,
    targets: T,
  ): this
  public implement(
    implementations: ImplementationOf<G>,
    targets?: Array<ConstructorOf<AbstractGrant>>,
  ): this {
    if (targets)
      this._grants
        .filter(grant =>
          targets.includes(grant.constructor as ConstructorOf<AbstractGrant>),
        )
        .forEach(instance => Object.assign(instance, implementations))
    else
      this._grants.forEach(instance => Object.assign(instance, implementations))

    return this
  }

  public async token(request: IRequest, response: IResponse): Promise<void> {
    const context = new Context(request, response)

    try {
      if (
        !(
          is(
            context.request.headers['content-type'] as string,
            'application/x-www-form-urlencoded',
          ) ||
          is(
            context.request.headers['Content-Type'] as string,
            'application/x-www-form-urlencoded',
          )
        )
      )
        throw new InvalidRequest(
          "Requests for an access token must be in the form 'application/x-www-form-urlencoded'",
          context,
        )

      const grantType = request.body.grant_type

      if (!grantType)
        throw new InvalidRequest('The grant type is missing', context)

      if (typeof grantType !== 'string')
        throw new InvalidRequest(
          `The grant type must be of type string but is set to ${typeof grantType}`,
          context,
        )

      // TODO: Invalid request or grant?
      if (!GRANT_TYPE_REGEXP.test(grantType))
        throw new InvalidRequest('The grant type is invalid', context)

      const grant = this._grants.find(grant => grant['type'] === grantType)

      if (!grant)
        throw new UnsupportedGrantType(
          'The grant type is not currently supported by the server.',
          context,
        )

      const token = await AbstractGrant.token.call(grant, context)

      const tokenType = this._tokenTypes.find(
        tokenType => tokenType.type === 'bearer',
      )

      if (!tokenType)
        throw new ServerError('No response type implemented on server', context)

      context.response.setHeader('Content-Type', 'application/json')
      context.response.status(200)
      context.response.send(JSON.stringify(tokenType.toJSON(token)))
    } catch (error: unknown) {
      if (isOAuth2Error(error)) {
        context.response.setHeader('Content-Type', 'application/json')
        context.response.status(error.status)
        context.response.send(
          JSON.stringify({
            error: error.error,
            error_description: error.description,
          }),
        )
      } else throw error
    }
  }

  public async release(request: IRequest, response: IResponse): Promise<void> {
    const context = new Context(request, response)
    const responseType =
      request.body.response_type ?? request.query.response_type

    if (!responseType)
      throw new InvalidRequest('The response type is missing', context)

    if (typeof responseType !== 'string')
      throw new InvalidRequest(
        `The response type must be of type string but is set to ${typeof responseType}`,
        context,
      )

    // TODO: Invalid request or grant?
    if (!RESPONSE_TYPE_REGEXP.test(responseType))
      throw new InvalidRequest('The response type is invalid', context)

    const flow = this._flows.find(flow => flow['type'] === responseType)

    if (!flow)
      throw new UnsupportedGrantType(
        'The response type is not currently supported by the server.',
        context,
      )

    await AbstractFlow.release.call(flow, context).catch(async error => {
      if (!isOAuth2Error(error)) throw error
      await AbstractFlow.releaseError.call(flow, error, context)
    })
  }

  public async authorize(
    exchanger: ConstructorOf<AbstractExchanger>,
    ticket: ITicket,
    user: IUser,
    request: IRequest,
    response: IResponse,
  ) {
    const context = new Context(request, response)

    const grant = this._grants.find(
      grant => grant instanceof exchanger,
    ) as Maybe<AbstractExchanger>

    if (!grant)
      throw new ServerError(
        `No instance of ${exchanger.name} grant found`,
        context,
      )

    const flow = grant['flow'] as Maybe<AbstractFlow>

    if (!flow)
      throw new ServerError(
        `The ${exchanger.name} grant has no ticket flow.`,
        context,
      )

    await AbstractFlow.authorize
      .call(flow, ticket, user, context)
      .catch(async error => {
        if (!isOAuth2Error(error)) throw error
        await AbstractFlow.authorizeError.call(
          flow,
          error,
          ticket,
          user,
          context,
        )
      })
  }

  public async deny(
    exchanger: ConstructorOf<AbstractExchanger>,
    ticket: ITicket,
    request: IRequest,
    response: IResponse,
  ) {
    const context = new Context(request, response)
    const grant = this._grants.find(
      grant => grant instanceof exchanger,
    ) as Maybe<AbstractExchanger>

    if (!grant)
      throw new ServerError(
        `No instance of ${exchanger.name} grant found`,
        context,
      )

    const flow = grant['flow'] as Maybe<AbstractFlow>

    if (!flow)
      throw new ServerError(
        `The ${exchanger.name} grant has no ticket flow.`,
        context,
      )

    await AbstractFlow.deny.call(flow, ticket, context).catch(async error => {
      if (!isOAuth2Error(error)) throw error
      await AbstractFlow.denyError.call(flow, error, ticket, context)
    })
  }
}
