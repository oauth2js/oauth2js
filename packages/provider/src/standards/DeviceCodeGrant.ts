/* eslint-disable accessor-pairs */
import {
  AbstractDeviceCodeFlow,
  AbstractDeviceCodeGrant,
} from '../abstracts/standards/AbstractDeviceCodeGrant'
import { ServerError } from '../errors/ServerError'
import { Maybe, PromiseToo } from '../helpers/types'
import { IClient } from '../interfaces/IClient'
import { IControllerImplementation } from '../interfaces/IControllerImplementation'
import { IDeviceCode } from '../interfaces/IDeviceCode'
import { IRequest } from '../interfaces/IRequest'
import { IScope } from '../interfaces/IScope'
import { ITicket } from '../interfaces/ITicket'
import { IToken } from '../interfaces/IToken'
import { IUser } from '../interfaces/IUser'
import { Context } from '../models/Context'
import {
  ClientCredentials,
  UseClientByCredentials,
} from '../traits/UseClientByCredentials'
import { UseClientById } from '../traits/UseClientById'

export class DeviceCodeFlow extends AbstractDeviceCodeFlow {
  protected selectUseClientProvider(
    context: Context,
  ): typeof UseClientByCredentials | typeof UseClientById {
    if (
      this.getClientByCredentials !==
      DeviceCodeFlow.prototype.getClientByCredentials
    )
      if (this.getClientById !== DeviceCodeFlow.prototype.getClientById)
        throw new ServerError(
          'Unable to resolve which client provider to use, conflict when `getClientByCredentials` and `getClientById` are set without a custom `selectUseClientProvider`.',
          context,
        )
      else return UseClientByCredentials

    if (this.getClientById !== DeviceCodeFlow.prototype.getClientById)
      return UseClientById

    throw new ServerError(
      'Unable to resolve which client provider to use, conflict when neither `getClientByCredentials` nor `getClientById` are defined.',
      context,
    )
  }

  protected getVerificationUri(
    deviceCode: IDeviceCode,
    context: Context,
  ): PromiseToo<string> {
    throw new ServerError(
      'The `getVerificationUri` definition is missing from the implementation.',
      context,
    )
  }

  protected getPullingInterval(context: Context): PromiseToo<number> {
    throw new ServerError(
      'The `getPullingInterval` definition is missing from the implementation.',
      context,
    )
  }

  protected saveDeviceCode(
    deviceCode: IDeviceCode,
    context: Context,
  ): PromiseToo<IDeviceCode> {
    throw new ServerError(
      'The `saveDeviceCode` definition is missing from the implementation.',
      context,
    )
  }

  protected authorizeDeviceCode(
    deviceCode: IDeviceCode,
    user: IUser,
    context: Context,
  ): PromiseToo<void> {
    throw new ServerError(
      'The `authorizeDeviceCode` definition is missing from the implementation.',
      context,
    )
  }

  protected denyDeviceCode(
    deviceCode: IDeviceCode,
    context: Context,
  ): PromiseToo<void> {
    throw new ServerError(
      'The `denyDeviceCode` definition is missing from the implementation.',
      context,
    )
  }

  protected getClientById(
    clientId: string,
    context: Context,
  ): PromiseToo<Maybe<IClient>> {
    throw new ServerError(
      'The `getClientById` definition is missing from the implementation.',
      context,
    )
  }

  protected getClientByCredentials(
    credentials: ClientCredentials,
    context: Context,
  ): PromiseToo<Maybe<IClient>> {
    throw new ServerError(
      'The `getClientByCredentials` definition is missing from the implementation.',
      context,
    )
  }

  protected getScopesObjects(
    scopes: string[],
    context: Context,
  ): PromiseToo<IScope[]> {
    throw new ServerError(
      'The `getScopesObjects` definition is missing from the implementation.',
      context,
    )
  }
}

export class DeviceCodeGrant extends AbstractDeviceCodeGrant {
  protected set pullingInterval(
    pullingInterval: number | ((context: Context) => PromiseToo<number>),
  ) {
    if (
      typeof pullingInterval !== 'function' &&
      typeof pullingInterval !== 'number'
    )
      throw new Error(
        `The 'pullingInterval' definition requires an implementation of type function or number, but is implemented as a ${typeof pullingInterval}`,
      )

    this.getPullingInterval = function (context) {
      if (typeof pullingInterval === 'function')
        return pullingInterval.call(this, context)
      return pullingInterval
    }
  }

  protected set verificationUri(
    verificationUri:
      | string
      | ((deviceCode: IDeviceCode, context: Context) => PromiseToo<string>),
  ) {
    if (
      typeof verificationUri !== 'function' &&
      typeof verificationUri !== 'string'
    )
      throw new Error(
        `The 'verificationUri' definition requires an implementation of type function or string, but is implemented as a ${typeof verificationUri}`,
      )

    this.getVerificationUri = function (
      deviceCode,
      context,
    ): PromiseToo<string> {
      if (typeof verificationUri === 'function')
        return verificationUri.call(this, deviceCode, context)
      return verificationUri
    }
  }

  protected set saveTokenByDeviceCode(
    saveTokenByDeviceCode: (
      token: IToken,
      deviceCode: IDeviceCode,
      context: Context,
    ) => PromiseToo<IToken>,
  ) {
    this.saveExchangeToken = saveTokenByDeviceCode
  }

  // @ts-expect-error: The `getPullingInterval` is shared between the grant and the flow.
  protected get getPullingInterval() {
    return this.flow['getPullingInterval'].bind(this.flow)
  }

  protected set getPullingInterval(getPullingInterval) {
    this.flow['getPullingInterval'] = getPullingInterval
  }

  protected set getVerificationUri(
    getVerificationUri: (
      deviceCode: IDeviceCode,
      context: Context,
    ) => PromiseToo<string>,
  ) {
    this.flow['getVerificationUri'] = getVerificationUri
  }

  protected set saveDeviceCode(
    saveDeviceCode: (
      deviceCode: IDeviceCode,
      context: Context,
    ) => PromiseToo<IDeviceCode>,
  ) {
    this.flow['saveDeviceCode'] = saveDeviceCode
  }

  protected set authorizeDeviceCode(
    authorizeDeviceCode: (
      deviceCode: IDeviceCode,
      user: IUser,
      context: Context,
    ) => PromiseToo<void>,
  ) {
    this.flow['authorizeDeviceCode'] = authorizeDeviceCode
  }

  protected set denyDeviceCode(
    denyDeviceCode: (
      deviceCode: IDeviceCode,
      context: Context,
    ) => PromiseToo<void>,
  ) {
    this.flow['denyDeviceCode'] = denyDeviceCode
  }

  protected set getClientById(
    getClientById: (
      clientId: string,
      context: Context,
    ) => PromiseToo<Maybe<IClient>>,
  ) {
    this.flow['getClientById'] = getClientById
  }

  protected set getClientByCredentials(
    getClientByCredentials: (
      credentials: ClientCredentials,
      context: Context,
    ) => PromiseToo<Maybe<IClient>>,
  ) {
    this.flow['getClientByCredentials'] = getClientByCredentials
  }

  protected set getScopesObjects(
    getScopesObjects: (
      scopes: string[],
      context: Context,
    ) => PromiseToo<IScope[]>,
  ) {
    this.flow['getScopesObjects'] = getScopesObjects
  }

  protected flow: DeviceCodeFlow = new DeviceCodeFlow()

  // @ts-expect-error: The `saveToken` must remain in the constructor implementation but for controller implementation issues it will be redirected to `saveExchangeToken`.
  protected override get saveToken() {
    return super.saveToken
  }

  protected override set saveToken(saveToken) {
    this.saveExchangeToken = async function (token, ticket, context) {
      return saveToken.call(this, token, context)
    }
  }

  protected getDeviceCodeByCode(
    code: string,
    context: Context,
  ): PromiseToo<Maybe<IDeviceCode>> {
    throw new ServerError(
      'The `getDeviceCodeByCode` definition is missing from the implementation.',
      context,
    )
  }

  protected getUserByDeviceCode(
    deviceCode: IDeviceCode,
    context: Context,
  ): PromiseToo<Maybe<IUser | false>> {
    throw new ServerError(
      'The `getUserByDeviceCode` definition is missing from the implementation.',
      context,
    )
  }

  protected getRequestId(
    request: IRequest,
    context: Context,
  ): PromiseToo<Maybe<string>>
  protected getRequestId(): PromiseToo<Maybe<string>> {
    return null
  }

  protected revokeDeviceCode(
    deviceCode: IDeviceCode,
    context: Context,
  ): PromiseToo<boolean> {
    throw new ServerError(
      'The `revokeDeviceCode` definition is missing from the implementation.',
      context,
    )
  }

  protected saveExchangeToken(
    token: IToken,
    ticket: ITicket,
    context: Context,
  ): PromiseToo<IToken> {
    throw new ServerError(
      'The `saveExchangeToken` definition is missing from the implementation.',
      context,
    )
  }
}

export interface DeviceCodeGrant
  extends IControllerImplementation<{
    pullingInterval: number | ((context: Context) => PromiseToo<number>)
    verificationUri:
      | string
      | ((deviceCode: IDeviceCode, context: Context) => PromiseToo<string>)
    revokeDeviceCode(
      deviceCode: IDeviceCode,
      context: Context,
    ): PromiseToo<boolean>
    saveDeviceCode(
      deviceCode: IDeviceCode,
      context: Context,
    ): PromiseToo<IDeviceCode>
    authorizeDeviceCode(
      deviceCode: IDeviceCode,
      user: IUser,
      context: Context,
    ): PromiseToo<void>
    denyDeviceCode(deviceCode: IDeviceCode, context: Context): PromiseToo<void>
    getClientById(
      clientId: string,
      context: Context,
    ): PromiseToo<Maybe<IClient>>
    getClientByCredentials(
      credentials: ClientCredentials,
      context: Context,
    ): PromiseToo<Maybe<IClient>>
    getUserByDeviceCode(
      deviceCode: IDeviceCode,
      context: Context,
    ): PromiseToo<Maybe<IUser | false>>
    getScopesObjects(scopes: string[], context: Context): PromiseToo<IScope[]>
    saveTokenByDeviceCode(
      token: IToken,
      deviceCode: IDeviceCode,
      context: Context,
    ): PromiseToo<IToken>
    saveToken(token: IToken, context: Context): PromiseToo<IToken>
    getRequestId(request: IRequest, context: Context): PromiseToo<Maybe<string>>
    getDeviceCodeByCode(
      code: string,
      context: Context,
    ): PromiseToo<Maybe<IDeviceCode>>
  }> {}
