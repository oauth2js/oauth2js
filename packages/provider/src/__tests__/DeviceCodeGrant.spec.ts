import { MockController } from '../../.jest/tools/Controller'
import { Database } from '../../.jest/tools/Database'
import {
  DatabaseClient,
  DatabaseDeviceCode,
  DatabaseScope,
  DatabaseUser,
} from '../../.jest/tools/DatabaseTypes'
import { DeviceCodeGrant } from '../standards/DeviceCodeGrant'

describe('Device Code Grant', () => {
  beforeEach(() => {
    jest.useFakeTimers()
  })

  afterEach(() => {
    jest.useRealTimers()
  })

  it('generate token', async () => {
    const db = Database()

    db.clients.insertOne({
      id: 'client-test-id',
      secret: 'client-test-secret',
      redirectUris: ['http://localhost:8080'],
    })
    db.users.insertOne({
      username: 'user-test-username',
      password: 'user-test-password',
    })
    db.scopes.insertOne({ alias: 'read', description: 'Read' })
    db.scopes.insertOne({ alias: 'write', description: 'Write' })

    const controller = MockController.from(controller =>
      controller.use(DeviceCodeGrant).implement({
        getClientByCredentials: credentials =>
          db.clients.findOne<DatabaseClient>(
            client =>
              client.id === credentials.id &&
              client.secret === credentials.secret,
          ),
        getScopesObjects: scopes =>
          db.scopes.find<DatabaseScope>(scope => scopes.includes(scope.alias)),
        saveDeviceCode: deviceCode => db.deviceCodes.insertOne(deviceCode),
        pullingInterval: 5 * 1000,
        verificationUri: 'http://localhost:8443',

        getRequestId: request => request.sessionID,

        getDeviceCodeByCode: code =>
          db.deviceCodes.findOne<DatabaseDeviceCode>(
            deviceCode => deviceCode.code === code,
          ),
        getUserByDeviceCode: (deviceCode: DatabaseDeviceCode) =>
          deviceCode.user ?? false,

        authorizeDeviceCode: ({ code }, user: DatabaseUser) =>
          void db.deviceCodes.updateOne<DatabaseDeviceCode>(
            { user },
            deviceCode => deviceCode.code === code,
          ),

        saveTokenByDeviceCode: (token, deviceCode) =>
          db.tokens.insertOne({ ...token, deviceCode }),

        revokeDeviceCode: ({ code }) =>
          !!db.deviceCodes.softDeleteOne(
            deviceCode => deviceCode.code === code,
          ),
      }),
    )

    let response = await controller.release({
      sessionID: 'testing',
      body: {
        response_type: 'device_code',
        client_id: 'client-test-id',
        client_secret: 'client-test-secret',
        scope: 'read+write',
      },
    })

    const body = JSON.parse(response.body)

    expect(response.status()).toBe(200)
    expect(response.headers).toMatchSnapshot()
    expect(body).toMatchSnapshot({
      device_code: expect.any(String),
      user_code: expect.any(String),
    })
    expect(db.deviceCodes.count()).toBe(1)
    expect(db.deviceCodes.findOne()).toMatchSnapshot({
      code: expect.any(String),
      codeExpiresAt: expect.any(Date),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      userCode: expect.any(String),
      client: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
      scopes: [
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
      ],
    })

    response = await controller.token({
      sessionID: 'testing',
      body: {
        grant_type: 'urn:ietf:params:oauth:grant-type:device_code',
        device_code: body.device_code,
        client_id: 'client-test-id',
      },
    })

    expect(response.status()).toBe(400)
    expect(response.headers).toMatchSnapshot()
    expect(JSON.parse(response.body)).toMatchSnapshot({
      error: 'authorization_pending',
    })

    response = await controller.token({
      sessionID: 'testing',
      body: {
        grant_type: 'urn:ietf:params:oauth:grant-type:device_code',
        device_code: body.device_code,
        client_id: 'client-test-id',
      },
    })

    expect(response.status()).toBe(400)
    expect(response.headers).toMatchSnapshot()
    expect(JSON.parse(response.body)).toMatchSnapshot({
      error: 'slow_down',
    })

    jest.setSystemTime(new Date(Date.now() + 5 * 1000))

    response = await controller.token({
      sessionID: 'testing',
      body: {
        grant_type: 'urn:ietf:params:oauth:grant-type:device_code',
        device_code: body.device_code,
        client_id: 'client-test-id',
      },
    })

    expect(response.status()).toBe(400)
    expect(response.headers).toMatchSnapshot()
    expect(JSON.parse(response.body)).toMatchSnapshot({
      error: 'authorization_pending',
    })

    response = await controller.authorize(
      DeviceCodeGrant,
      db.deviceCodes.findOneOrFail(),
      db.users.findOneOrFail(),
    )

    expect(response.status()).toBe(204)
    expect(response.headers).toMatchSnapshot()
    expect(response.body).toBeUndefined()

    expect(db.deviceCodes.count()).toBe(1)
    expect(db.deviceCodes.findOne()).toMatchSnapshot({
      code: expect.any(String),
      codeExpiresAt: expect.any(Date),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      userCode: expect.any(String),
      client: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
      scopes: [
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
      ],
      user: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
    })

    response = await controller.token({
      sessionID: 'testing',
      body: {
        grant_type: 'urn:ietf:params:oauth:grant-type:device_code',
        device_code: body.device_code,
        client_id: 'client-test-id',
      },
    })

    expect(response.status()).toBe(400)
    expect(response.headers).toMatchSnapshot()
    expect(JSON.parse(response.body)).toMatchSnapshot({
      error: 'slow_down',
    })

    jest.setSystemTime(new Date(Date.now() + 5 * 1000))

    response = await controller.token({
      sessionID: 'testing',
      body: {
        grant_type: 'urn:ietf:params:oauth:grant-type:device_code',
        device_code: body.device_code,
        client_id: 'client-test-id',
      },
    })

    expect(response.status()).toBe(200)
    expect(response.headers).toMatchSnapshot()
    expect(JSON.parse(response.body)).toMatchSnapshot({
      access_token: expect.any(String),
      refresh_token: expect.any(String),
    })
    expect(db.deviceCodes.count({ includeDeleted: true })).toBe(1)
    expect(
      db.deviceCodes.findOne(void 0, { includeDeleted: true }),
    ).toMatchSnapshot({
      code: expect.any(String),
      codeExpiresAt: expect.any(Date),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      deletedAt: expect.any(Date),
      userCode: expect.any(String),
      client: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
      scopes: [
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
        {
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        },
      ],
      user: {
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      },
    })
  })
})

declare global {
  namespace OAuth2JS {
    namespace Server {
      interface Request {
        sessionID: string
      }
    }
  }
}
