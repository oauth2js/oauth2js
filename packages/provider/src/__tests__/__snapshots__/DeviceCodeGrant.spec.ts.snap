// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Device Code Grant generate token 1`] = `
Object {
  "Content-Type": "application/json",
}
`;

exports[`Device Code Grant generate token 2`] = `
Object {
  "device_code": Any<String>,
  "expires_in": 300,
  "interval": 5000,
  "user_code": Any<String>,
  "verification_uri": "http://localhost:8443",
  "verification_uri_complete": null,
}
`;

exports[`Device Code Grant generate token 3`] = `
Object {
  "client": Object {
    "createdAt": Any<ClockDate>,
    "deletedAt": null,
    "id": "client-test-id",
    "redirectUris": Array [
      "http://localhost:8080",
    ],
    "secret": "client-test-secret",
    "updatedAt": Any<ClockDate>,
  },
  "code": Any<String>,
  "codeExpiresAt": Any<ClockDate>,
  "createdAt": Any<ClockDate>,
  "deletedAt": null,
  "scopes": Array [
    Object {
      "alias": "read",
      "createdAt": Any<ClockDate>,
      "deletedAt": null,
      "description": "Read",
      "updatedAt": Any<ClockDate>,
    },
    Object {
      "alias": "write",
      "createdAt": Any<ClockDate>,
      "deletedAt": null,
      "description": "Write",
      "updatedAt": Any<ClockDate>,
    },
  ],
  "updatedAt": Any<ClockDate>,
  "userCode": Any<String>,
}
`;

exports[`Device Code Grant generate token 4`] = `
Object {
  "Content-Type": "application/json",
}
`;

exports[`Device Code Grant generate token 5`] = `
Object {
  "error": "authorization_pending",
  "error_description": "Authorization is pending",
}
`;

exports[`Device Code Grant generate token 6`] = `
Object {
  "Content-Type": "application/json",
}
`;

exports[`Device Code Grant generate token 7`] = `
Object {
  "error": "slow_down",
  "error_description": "Slow down",
}
`;

exports[`Device Code Grant generate token 8`] = `
Object {
  "Content-Type": "application/json",
}
`;

exports[`Device Code Grant generate token 9`] = `
Object {
  "error": "authorization_pending",
  "error_description": "Authorization is pending",
}
`;

exports[`Device Code Grant generate token 10`] = `Object {}`;

exports[`Device Code Grant generate token 11`] = `
Object {
  "client": Object {
    "createdAt": Any<ClockDate>,
    "deletedAt": null,
    "id": "client-test-id",
    "redirectUris": Array [
      "http://localhost:8080",
    ],
    "secret": "client-test-secret",
    "updatedAt": Any<ClockDate>,
  },
  "code": Any<String>,
  "codeExpiresAt": Any<ClockDate>,
  "createdAt": Any<ClockDate>,
  "deletedAt": null,
  "scopes": Array [
    Object {
      "alias": "read",
      "createdAt": Any<ClockDate>,
      "deletedAt": null,
      "description": "Read",
      "updatedAt": Any<ClockDate>,
    },
    Object {
      "alias": "write",
      "createdAt": Any<ClockDate>,
      "deletedAt": null,
      "description": "Write",
      "updatedAt": Any<ClockDate>,
    },
  ],
  "updatedAt": Any<ClockDate>,
  "user": Object {
    "createdAt": Any<ClockDate>,
    "deletedAt": null,
    "password": "user-test-password",
    "updatedAt": Any<ClockDate>,
    "username": "user-test-username",
  },
  "userCode": Any<String>,
}
`;

exports[`Device Code Grant generate token 12`] = `
Object {
  "Content-Type": "application/json",
}
`;

exports[`Device Code Grant generate token 13`] = `
Object {
  "error": "slow_down",
  "error_description": "Slow down",
}
`;

exports[`Device Code Grant generate token 14`] = `
Object {
  "Content-Type": "application/json",
}
`;

exports[`Device Code Grant generate token 15`] = `
Object {
  "access_token": Any<String>,
  "expires_in": 7200,
  "refresh_token": Any<String>,
  "token_type": "Bearer",
}
`;

exports[`Device Code Grant generate token 16`] = `
Object {
  "client": Object {
    "createdAt": Any<ClockDate>,
    "deletedAt": null,
    "id": "client-test-id",
    "redirectUris": Array [
      "http://localhost:8080",
    ],
    "secret": "client-test-secret",
    "updatedAt": Any<ClockDate>,
  },
  "code": Any<String>,
  "codeExpiresAt": Any<ClockDate>,
  "createdAt": Any<ClockDate>,
  "deletedAt": Any<ClockDate>,
  "scopes": Array [
    Object {
      "alias": "read",
      "createdAt": Any<ClockDate>,
      "deletedAt": null,
      "description": "Read",
      "updatedAt": Any<ClockDate>,
    },
    Object {
      "alias": "write",
      "createdAt": Any<ClockDate>,
      "deletedAt": null,
      "description": "Write",
      "updatedAt": Any<ClockDate>,
    },
  ],
  "updatedAt": Any<ClockDate>,
  "user": Object {
    "createdAt": Any<ClockDate>,
    "deletedAt": null,
    "password": "user-test-password",
    "updatedAt": Any<ClockDate>,
    "username": "user-test-username",
  },
  "userCode": Any<String>,
}
`;
